packages/network-controller/src/NetworkController.ts

```diff
@@ -991,201 +1293,627 @@ export class NetworkController extends BaseController<
+      throw new Error(
+        `Cannot update network: \`defaultRpcEndpointUrl\` '${newInfuraRpcEndpointUrl}' must match an entry in \`rpcEndpoints\` (${inspect(
+          newRpcEndpointUrls,
+        )})`,
+      );
+    }
+
+    const operations: {
+      type: 'add' | 'remove' | 'noop';
+      value: RpcEndpoint;
+    }[] = [];
+    if (newChainId === existingChainId) {
+      for (const existingRpcEndpoint of existingNetworkConfiguration.rpcEndpoints) {
+        if (
+          !setOfNewRpcEndpointFields.some(
+            (newRpcEndpointFields) =>
+              newRpcEndpointFields.networkClientId ===
+                existingRpcEndpoint.networkClientId &&
+              newRpcEndpointFields.type === existingRpcEndpoint.type &&
+              newRpcEndpointFields.url === existingRpcEndpoint.url,
+          )
+        ) {
+          operations.push({
+            type: 'remove',
+            value: existingRpcEndpoint,
+          });
+        }
+      }
+      for (const newRpcEndpointFields of setOfNewRpcEndpointFields) {
+        const existingRpcEndpoint =
+          existingNetworkConfiguration.rpcEndpoints.find(
+            (rpcEndpoint) =>
+              rpcEndpoint.networkClientId ===
+                newRpcEndpointFields.networkClientId &&
+              rpcEndpoint.type === newRpcEndpointFields.type &&
+              rpcEndpoint.url === newRpcEndpointFields.url,
+          );
+        if (existingRpcEndpoint === undefined) {
+          const newRpcEndpoint =
+            newRpcEndpointFields.type === RpcEndpointType.Infura
+              ? newRpcEndpointFields
+              : { ...newRpcEndpointFields, networkClientId: uuidV4() };
+          operations.push({
+            type: 'add',
+            value: newRpcEndpoint,
+          });
+        } else {
+          operations.push({
+            type: 'noop',
+            value: existingRpcEndpoint,
+          });
+        }
+      }
+    } else {
+      for (const existingRpcEndpoint of existingNetworkConfiguration.rpcEndpoints) {
+        operations.push({
+          type: 'remove',
+          value: existingRpcEndpoint,
+        });
+      }
+      for (const newRpcEndpointFields of setOfNewRpcEndpointFields) {
+        const newRpcEndpoint =
+          newRpcEndpointFields.type === RpcEndpointType.Infura
+            ? newRpcEndpointFields
+            : { ...newRpcEndpointFields, networkClientId: uuidV4() };
+        operations.push({
+          type: 'add',
+          value: newRpcEndpoint,
+        });
+      }
+    }
+
+    const newRpcEndpoints: RpcEndpoint[] = [];
+    for (const operation of operations) {
+      if (operation.type === 'remove') {
+        const networkClient = this.getNetworkClientById(
+          operation.value.networkClientId,
+        );
+        networkClient.destroy();
+        delete autoManagedNetworkClientRegistry[
+          networkClient.configuration.type
+        ][operation.value.networkClientId];
```

Comment:
This and other `delete` operations here will probably trigger the ["prototype-polluting assignment"](https://github.com/MetaMask/core/pull/4041/files#diff-78fe44eeec5755b58fbeef2367e9a09861f63ba1d71dacd6494060d609187e76L110-L129) CodeQL alert. `isSafeDynamicKey`

```suggestion
        if (![networkClient.configuration.type, operation.value.networkClientId].every((e) => isSafeDynamicKey(e))) {
          continue;
        }
        delete autoManagedNetworkClientRegistry[
          networkClient.configuration.type
        ][operation.value.networkClientId];
```

Reply:
I believe this only applies if the key you're using comes from the outside, right? But in this case I'm controlling what `networkClient.configuration.type` and `operation.value.networkClientId` can be. So perhaps this isn't needed?

Reply:
I think the static code analysis just flags any dynamic string being used for property access, which should apply here. But I don't think there's any actual risk of a pollution/injection attack here. We can always come back and add `isSafeDynamicKey` if this is flagged.